File: src/mongo/schema.ts
import { ObjectId, Binary, Long, Decimal128, UUID } from "mongodb";
import type { MongoClient, Document } from "mongodb";
import { getTimeout } from "../lib/timeout.ts";
import { validateCollectionExists } from "./collections.ts";

type FieldInfo = {
  path: string;
  types: string[];
  presence: number;
};

export type SchemaResult = {
  database: string;
  collection: string;
  sampleSize: number;
  totalDocuments: number;
  fields: FieldInfo[];
};

type SchemaOpts = {
  dbName: string;
  collName: string;
  sampleSize: number;
  maxDepth?: number;
};

export async function inferSchema(client: MongoClient, opts: SchemaOpts): Promise<SchemaResult> {
  await validateCollectionExists(client, { dbName: opts.dbName, collName: opts.collName });

  const timeout = getTimeout();
  const collection = client.db(opts.dbName).collection(opts.collName);

  const totalDocuments = await collection.estimatedDocumentCount({ maxTimeMS: timeout });

  const effectiveSize = Math.min(opts.sampleSize, totalDocuments || opts.sampleSize);
  const docs = await collection
    .aggregate<Document>([{ $sample: { size: effectiveSize } }], { maxTimeMS: timeout })
    .toArray();

  const fieldMap = new Map<string, { types: Set<string>; count: number }>();

  for (const doc of docs) {
    const seen = new Set<string>();
    walkDocument({ doc, prefix: "", seen, fields: fieldMap, depth: 1, maxDepth: opts.maxDepth });
  }

  const fields: FieldInfo[] = Array.from(fieldMap.entries())
    .map(([path, info]) => ({
      path,
      types: Array.from(info.types).sort(),
      presence: docs.length > 0 ? Math.round((info.count / docs.length) * 100) / 100 : 0,
    }))
    .sort((a, b) => a.path.localeCompare(b.path));

  return {
    database: opts.dbName,
    collection: opts.collName,
    sampleSize: docs.length,
    totalDocuments,
    fields,
  };
}

type WalkOpts = {
  doc: Document;
  prefix: string;
  seen: Set<string>;
  fields: Map<string, { types: Set<string>; count: number }>;
  depth: number;
  maxDepth?: number;
};

function walkDocument({ doc, prefix, seen, fields, depth, maxDepth }: WalkOpts): void {
  for (const [key, value] of Object.entries(doc)) {
    const path = prefix ? `${prefix}.${key}` : key;
    recordField({ path, typeName: getTypeName(value), seen, fields });

    if (maxDepth !== undefined && depth >= maxDepth) {
      continue;
    }

    if (isPlainObject(value)) {
      walkDocument({ doc: value as Document, prefix: path, seen, fields, depth: depth + 1, maxDepth });
    } else if (Array.isArray(value)) {
      walkArrayElements({ arr: value, parentPath: path, seen, fields, depth, maxDepth });
    }
  }
}

type WalkArrayOpts = {
  arr: unknown[];
  parentPath: string;
  seen: Set<string>;
  fields: Map<string, { types: Set<string>; count: number }>;
  depth: number;
  maxDepth?: number;
};

function walkArrayElements({ arr, parentPath, seen, fields, depth, maxDepth }: WalkArrayOpts): void {
  const elemPath = `${parentPath}.$`;
  for (const elem of arr) {
    recordFieldType({ path: elemPath, typeName: getTypeName(elem), fields });

    if (maxDepth === undefined || depth < maxDepth) {
      if (isPlainObject(elem)) {
        walkDocument({
          doc: elem as Document,
          prefix: elemPath,
          seen,
          fields,
          depth: depth + 1,
          maxDepth,
        });
      }
    }
  }

  if (arr.length > 0 && !seen.has(elemPath)) {
    seen.add(elemPath);
    const field = fields.get(elemPath);
    if (field) {
      field.count++;
    }
  }
}

type RecordFieldOpts = {
  path: string;
  typeName: string;
  seen: Set<string>;
  fields: Map<string, { types: Set<string>; count: number }>;
};

function recordField({ path, typeName, seen, fields }: RecordFieldOpts): void {
  recordFieldType({ path, typeName, fields });
  if (!seen.has(path)) {
    seen.add(path);
    const field = fields.get(path);
    if (field) {
      field.count++;
    }
  }
}

type RecordFieldTypeOpts = {
  path: string;
  typeName: string;
  fields: Map<string, { types: Set<string>; count: number }>;
};

function recordFieldType({ path, typeName, fields }: RecordFieldTypeOpts): void {
  let field = fields.get(path);
  if (!field) {
    field = { types: new Set(), count: 0 };
    fields.set(path, field);
  }
  field.types.add(typeName);
}

function getTypeName(value: unknown): string {
  if (value === null) {
    return "null";
  }
  if (value === undefined) {
    return "null";
  }
  if (value instanceof ObjectId) {
    return "ObjectId";
  }
  if (value instanceof Date) {
    return "date";
  }
  if (value instanceof Binary) {
    return "binary";
  }
  if (value instanceof Long) {
    return "long";
  }
  if (value instanceof Decimal128) {
    return "decimal";
  }
  if (value instanceof UUID) {
    return "uuid";
  }
  if (value instanceof RegExp) {
    return "regex";
  }
  if (Array.isArray(value)) {
    return "array";
  }
  if (typeof value === "string") {
    return "string";
  }
  if (typeof value === "number") {
    return Number.isInteger(value) ? "int" : "double";
  }
  if (typeof value === "boolean") {
    return "boolean";
  }
  if (typeof value === "object") {
    return "object";
  }
  return typeof value;
}

function isPlainObject(value: unknown): boolean {
  if (value === null || value === undefined) {
    return false;
  }
  if (typeof value !== "object") {
    return false;
  }
  if (Array.isArray(value)) {
    return false;
  }
  if (value instanceof ObjectId) {
    return false;
  }
  if (value instanceof Date) {
    return false;
  }
  if (value instanceof Binary) {
    return false;
  }
  if (value instanceof Long) {
    return false;
  }
  if (value instanceof Decimal128) {
    return false;
  }
  if (value instanceof UUID) {
    return false;
  }
  if (value instanceof RegExp) {
    return false;
  }
  return true;
}

File: internal/mongo/schema.go
package mongo

import (
	"context"
	"math"
	"sort"
	"time"

	"go.mongodb.org/mongo-driver/v2/bson"
)

type FieldInfo struct {
	Path     string   `json:"path"`
	Types    []string `json:"types"`
	Presence float64  `json:"presence"`
}

type SchemaResult struct {
	Database       string
	Collection     string
	SampleSize     int
	TotalDocuments int64
	Fields         []FieldInfo
}

type SchemaOpts struct {
	Ref
	SampleSize int
	MaxDepth   int // 0 = unlimited
}

type fieldAgg struct {
	types map[string]bool
	count int
}

type walker struct {
	fields   map[string]*fieldAgg
	seen     map[string]bool
	maxDepth int
}

// InferSchema samples documents and aggregates field paths, types, and
// presence rates. Array element types are recorded under "path.$".
func (s *Session) InferSchema(ctx context.Context, opts SchemaOpts) (SchemaResult, error) {
	if err := s.ValidateCollectionExists(ctx, opts.Ref); err != nil {
		return SchemaResult{}, err
	}

	collection := s.Client.Database(opts.DB).Collection(opts.Collection)
	totalDocuments, err := collection.EstimatedDocumentCount(ctx)
	if err != nil {
		return SchemaResult{}, err
	}

	effectiveSize := opts.SampleSize
	if totalDocuments > 0 && int64(effectiveSize) > totalDocuments {
		effectiveSize = int(totalDocuments)
	}

	cursor, err := collection.Aggregate(ctx, bson.A{
		bson.D{{Key: "$sample", Value: bson.D{{Key: "size", Value: effectiveSize}}}},
	})
	if err != nil {
		return SchemaResult{}, err
	}
	var docs []bson.D
	if err := cursor.All(ctx, &docs); err != nil {
		return SchemaResult{}, err
	}

	return SchemaResult{
		Database:       opts.DB,
		Collection:     opts.Collection,
		SampleSize:     len(docs),
		TotalDocuments: totalDocuments,
		Fields:         InferFields(docs, opts.MaxDepth),
	}, nil
}

// InferFields aggregates field paths, types, and presence rates from sampled
// documents — the pure core of schema inference.
func InferFields(docs []bson.D, maxDepth int) []FieldInfo {
	w := &walker{fields: map[string]*fieldAgg{}, maxDepth: maxDepth}
	for _, doc := range docs {
		w.seen = map[string]bool{}
		w.walkDocument(doc, "", 1)
	}

	fields := make([]FieldInfo, 0, len(w.fields))
	for path, agg := range w.fields {
		types := make([]string, 0, len(agg.types))
		for t := range agg.types {
			types = append(types, t)
		}
		sort.Strings(types)
		presence := 0.0
		if len(docs) > 0 {
			presence = math.Round(float64(agg.count)/float64(len(docs))*100) / 100
		}
		fields = append(fields, FieldInfo{Path: path, Types: types, Presence: presence})
	}
	sort.Slice(fields, func(i, j int) bool { return fields[i].Path < fields[j].Path })
	return fields
}

func (w *walker) walkDocument(doc bson.D, prefix string, depth int) {
	for _, elem := range doc {
		path := elem.Key
		if prefix != "" {
			path = prefix + "." + elem.Key
		}
		w.recordField(path, typeName(elem.Value))

		if w.maxDepth > 0 && depth >= w.maxDepth {
			continue
		}

		switch v := elem.Value.(type) {
		case bson.D:
			w.walkDocument(v, path, depth+1)
		case bson.A:
			w.walkArrayElements(v, path, depth)
		}
	}
}

func (w *walker) walkArrayElements(arr bson.A, parentPath string, depth int) {
	elemPath := parentPath + ".$"
	for _, elem := range arr {
		w.recordFieldType(elemPath, typeName(elem))

		if w.maxDepth == 0 || depth < w.maxDepth {
			if doc, ok := elem.(bson.D); ok {
				w.walkDocument(doc, elemPath, depth+1)
			}
		}
	}

	if len(arr) > 0 && !w.seen[elemPath] {
		w.seen[elemPath] = true
		if field, ok := w.fields[elemPath]; ok {
			field.count++
		}
	}
}

func (w *walker) recordField(path, tn string) {
	w.recordFieldType(path, tn)
	if !w.seen[path] {
		w.seen[path] = true
		if field, ok := w.fields[path]; ok {
			field.count++
		}
	}
}

func (w *walker) recordFieldType(path, tn string) {
	field, ok := w.fields[path]
	if !ok {
		field = &fieldAgg{types: map[string]bool{}}
		w.fields[path] = field
	}
	field.types[tn] = true
}

const maxSafeInteger = 1<<53 - 1

func typeName(value any) string {
	switch v := value.(type) {
	case nil, bson.Null, bson.Undefined:
		return "null"
	case bson.ObjectID:
		return "ObjectId"
	case bson.DateTime, time.Time:
		return "date"
	case bson.Binary:
		if v.Subtype == bson.TypeBinaryUUID {
			return "uuid"
		}
		return "binary"
	case bson.Decimal128:
		return "decimal"
	case bson.Regex:
		return "regex"
	case bson.A, []any:
		return "array"
	case string:
		return "string"
	case int32:
		return "int"
	case int64:
		// The TS driver promotes safe int64s to JS numbers, which the TS
		// schema reports as "int"; only unsafe magnitudes surface as "long".
		if v >= -maxSafeInteger && v <= maxSafeInteger {
			return "int"
		}
		return "long"
	case float64:
		if v == math.Trunc(v) && !math.IsInf(v, 0) && !math.IsNaN(v) {
			return "int"
		}
		return "double"
	case bool:
		return "boolean"
	case bson.D, bson.M, map[string]any:
		return "object"
	default:
		return "object"
	}
}
